File: c++/2023/solutions/day06-2.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <cmath>
#include <algorithm>

int main() {
	std::ifstream input{ "../inputs/2023/day06-input.txt" };
	std::string times, distance;
	std::getline(input, times);
	std::getline(input, distance);
	
	times = times.substr(10);
	distance = distance.substr(10);
	times.erase(std::remove(times.begin(), times.end(), ' '), times.end());
	distance.erase(std::remove(distance.begin(), distance.end(), ' '), distance.end());
	long long t = std::stoll(times);
	long long r = std::stoll(distance);
	
	double dis = std::sqrt(t*t - 4*r);
	long long smallest = std::ceil((t - dis) / 2);
	long long biggest = std::floor((t + dis) / 2);
	
	std::cout << (biggest - smallest) + 1 << '\n';
}

File: haskell/2023/solutions/day06-2.hs
import Data.List as L

-- T = total time, H = held time, D = total distance, R = record distance
-- H^2 - TH + R < 0
-- (T +/- sqrt(T^2 - 4*R))/2

winWays :: [Integer] -> Integer
winWays info = floor ((t + disc) / 2) - ceiling ((t - disc) / 2) + 1
   where t = fromIntegral (info !! 0) :: Double
         r = fromIntegral (info !! 1)
         disc = sqrt (t^2 - 4*r)

solution :: String -> String
solution x = show $ winWays [nums !! 0, nums !! 1]
    where nums = map (read . concat . drop 1 . words) (lines x) :: [Integer]

main = do
       file <- readFile "../inputs/2023/day06-input.txt"
       putStrLn $ solution file